Implement auto-renaming of multiple selected files

jeremypw opened the pull request:
Previous developments squashed into one commit
* Show bulk rename dialog when more than one file selected for renaming
* Allow prefixes and suffixes
* Allow basename same as original, fixed text or partial replacement of original
* Simplified default options
* Allow editing of modifiers if required
* Allow sorting and reversing number sequences
* Allow partial renaming of list

Possible future enhancements in other PRs
* Other file attributes as modifiers or sortby
* Metadata as modifiers
* Character sequence as modifier
* Reorder modifiers

jeremypw commented:
Initial Dialog:

![BulkRenamerOpeningWindow](https://user-images.githubusercontent.com/10513844/151555322-717f104e-09fb-4a98-98bd-68cd71e337f0.png)

jeremypw commented:
Fixed Basename:
![BulkRenamerFixedBasename](https://user-images.githubusercontent.com/10513844/151585451-e315665f-c737-45a4-ae80-b63550a4335d.png)
Original with replacement Basename:
![BulkRenamerReplace](https://user-images.githubusercontent.com/10513844/151585518-f2a47ed6-64c2-4d9b-9962-3aa4716ab124.png)
Modifier choice:
![BulkRenamerModifierOptions](https://user-images.githubusercontent.com/10513844/151585625-5d0be337-3a11-433b-8dd6-919da87f25ab.png)
Number Sequence options:
![BulkRenamerNumberSeqEdit](https://user-images.githubusercontent.com/10513844/151585562-31379681-113b-497d-a71a-92eff219852c.png)
Date Options:
![BulkRenamerDateEdit](https://user-images.githubusercontent.com/10513844/151585675-d9b2d23d-3df1-40bd-bc4f-3ff9e8d6cf97.png)
Fixed text options:
![BulkRenamerTextEdit](https://user-images.githubusercontent.com/10513844/151585736-03617a3c-bd0a-429f-953e-4a1821518f72.png)

danrabbit commented:
This is definitely an improvement from the previous branch!

I can see where you're going with laying out the prefix and suffix stuff like this, but I think mixing actions with the output is a little confusing and it means actions move around as the UI becomes more complicated and eventually the renaming method becomes lost in all the buttons.

You can get into some situations here that probably don't make sense adding the same prefixes and suffixes that have already been used:
![Screenshot from 2022-02-02 15 19 54](https://user-images.githubusercontent.com/7277719/152254113-dea09df6-be0c-4b92-911a-ad5a7995a668.png)

I think another problem here is that you can make a window that's larger than the screen this way pretty easily on a small display

I'm having a hard time thinking of a better way to solve it than the design that GNOME is using. Having a single text entry where you add placeholders like "[Original file name]" or "[1,2,3]" seems incredibly flexible and with a much simpler UI without some of these other issues. It would be very easy to make something into a suffix or prefix or add arbitrary text without having to have explicit actions for that in the UI other than just adding new placeholders for discoverability.

jeremypw commented:
I was trying not to restrict the user too much but that means allowing to them to do something not sensible (as long as it is not destructive).  I can certainly restrict some things like having more than one prefix/suffix of the same kind and/or the total number of modifications.  Could maybe restyle the "Add -fix" tool to distinguish more obviously from the modifiers?  Could have a single separate the "Add -fix" tool like Gnome which defaults to adding a suffix and have an extra option in the modifier menu to make it a prefix?

Personally, I did not find the Gnome interface particularly flexible (at least not obviously).  It only offers to add a number range as a suffix for me (with few options).  I cannot edit placeholders and if I type one in as a prefix it is not interpreted.  Maybe it has improved in a later version?

jeremypw commented:
It seems the more advanced abilities of Gnome Files depend on having Tracker installed and working.  I tried installing it but it did not seem to make any difference.  Maybe working better in elementary 7?  I am still using elementary 6.1

jeremypw commented:
The latest commits limit the modifications to one suffix and/or one prefix (max number is parameterized). The modifier buttons' labels now reflect what they do more precisely.

![RenamerLimitedAffixes](https://user-images.githubusercontent.com/10513844/152570886-439dfe6a-33f3-4c7e-a43f-947e6cec730d.png)

jeremypw commented:
The window size still tends to change so I may push an alternate design to a different PR.

jeremypw commented:
Using a set of radiobuttons for sorting choice increases the minimum window width and makes width changing on adding modifiers less likely.

![RenamerRadioButtonSortby](https://user-images.githubusercontent.com/10513844/152577325-688973ff-fca0-4e6c-a211-7b6383c65e28.png)

jeremypw commented:
Replaced basename combo with radiobuttons:

Unchanged basename (default)
![BulkRenamerNoCombo](https://user-images.githubusercontent.com/10513844/153055891-8edf2dcc-1074-4a66-ab00-d115a4e5f425.png)

Modified basename 
![BulkRenamerNoComboModified](https://user-images.githubusercontent.com/10513844/153056309-675e494e-d0c4-44f0-8d58-102bb03127e2.png)

New fixed basename
![BulkRenamerNoComboNew](https://user-images.githubusercontent.com/10513844/153056328-2a7655ad-916e-4bb5-9915-55b191fd1aaa.png)

jeremypw commented:
@danrabbit I have pared down the interface some more and replaced the combobox with radiobuttons.  Is this going in the right direction?

jeremypw commented:
Sortby radiobuttons shown when number sequence chosen
![BulkRenamerNoComboNumSeq](https://user-images.githubusercontent.com/10513844/153057031-08a20af9-457e-417f-9586-47da42ba9762.png)

jeremypw commented:
@danirabbit Should I continue with this? It has undergone several design iterations.  Maybe time to squash and resubmit?

marco4mp commented:
Hi, sorry for asking here. But, is there any progress on this? It's a pretty useful feature.

jeremypw commented:
Sorry Marco,  still awaiting approval.  It is hard to get large amounts of
new code approved but in this case there is no way of breaking it up into
incremental commits.

I could try reminding Danielle again but she is focused on other projects
at the moment.

Jeremy

On Thu, 13 Apr 2023 at 23:54, Marco M. ***@***.***> wrote:

> Hi, sorry for asking here. But, is there any progress on this? It's a
> pretty useful feature.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/elementary/files/pull/1999#issuecomment-1507702415>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQG3NBFH76FK6P5XZHCS4DXBB7YXANCNFSM5NAXK62A>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

peteruithoven commented:
Great work! Would love to see something like this integrated into elementary OS. 

Some quick UX feedback:
People often recommend against horizontal radio buttons:
https://uxplanet.org/radio-buttons-ux-design-588e5c0a50dc#9b5b
For some cases it might be interesting to use the ModeButton? 
![Screenshot from 2023-04-15 14 29 26@2x](https://user-images.githubusercontent.com/523210/232223637-3244c4bb-24eb-4108-b260-b9726e6d5133.png)
![Screenshot from 2023-04-15 14 54 34@2x](https://user-images.githubusercontent.com/523210/232224913-dcf31405-6211-4b9a-9211-b1584a125eaf.png)
![Screenshot from 2023-04-15 14 53 54@2x](https://user-images.githubusercontent.com/523210/232224915-0e723831-f726-4f31-b880-7f77b7fb2558.png)
![Screenshot from 2023-04-15 14 53 34@2x](https://user-images.githubusercontent.com/523210/232224916-eacc0f2e-6275-497e-8cac-2f00e0c778bf.png)

"The number in order of" might be confused by a kind of sorting/filter option, since it's directly above the before & after preview list.
Why not integrate this "The number in order of" in the Number sequence dialog? 

You must have considered this but why not configure the name parts in a vertical list? I understand it's slightly less intuitive, but it makes the layout easier / more scalable. It might even make it possible to move/drag parts around. 
It could function a bit like the lists you see under Firewall and custom keyboard shortcuts? Having some quick options when you add something, and adding a dropdown per item to customize them
![Screenshot from 2023-04-15 14 49 08@2x](https://user-images.githubusercontent.com/523210/232224670-2d2a1b50-7335-43af-ac2d-01de4496d532.png)
![Screenshot from 2023-04-15 14 48 28@2x](https://user-images.githubusercontent.com/523210/232224731-489358e4-c7d6-4f8b-94c4-a2c758af4a8a.png)
I've not seen something where you can change the order though, which is a big problem for this usecase.
But  then again, if the max 1 prefix and 1 postfix is enough there is no need. Keeping it simple & constrained might be the way to go to get the first version in.

jeremypw commented:
@peteruithoven Thanks very much for your feedback :smile:  I am quite happy to iterate on the design as I know that is not my strong point. 

> For some cases it might be interesting to use the ModeButton?

Yes, good idea - I'll try it.


> Why not integrate this "The number in order of" in the Number sequence dialog?

Good idea! If there was a reason I didn't do it this way I've forgotten now :stuck_out_tongue: 

> You must have considered this but why not configure the name parts in a vertical list? 

One of my earlier efforts did have a more vertical layout (see https://github.com/jeremypw/bulk-renamer-plugin) but I liked that the horizontal parts matched the actual name so seems more intuitive.  If width were an issue could probably make the layout responsive. Supporting narrow layouts was probably not a thing when I started this. If we were to want to support arbitrary numbers of prefixes and suffixes then the UI would have to be rethought anyway.

> Keeping it simple & constrained might be the way to go to get the first version in.

Absolutely!  Not sure what the minimum feature set that would be useful for a start is here, but suggestions for paring it down further welcome.

danirabbit commented on plugins/meson.build:
```diff
@@ -4,3 +4,4 @@ subdir('pantheon-files-ctags')
 subdir('pantheon-files-cloud')
 subdir('git')
 subdir('send-by-email')
+# subdir('bulk-renamer')
```
leftover from previous commits?

jeremypw replied:
Fixed

danirabbit reviewed (approved):
I'm not super in love with this design, but it is a hotly requested feature so I'm inclined to merge and iterate on it later :+1:

jeremypw commented:
Yes, I wrote this a while back and if I went through it again I would definitely find many ways it could be improved in both coding and design but at least it is something that works and can be improved upon in the future depending on feedback.

danirabbit reviewed (approved):
:shipit:

jeremypw commented:
@peteruithoven As this has now been merged, could you raise issues against specific design flaws so that we can iterate on it? Thanks.
